DRA: ResourceSlice informer

pohly opened the pull request:
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:

The original idea was that avoiding repeated conversion of ResourceSlices to the internal type by doing it in an informer would bring some performance advantage. Because the DRA scheduler_perf benchmarks didn't clearly distinguish between "setup phase" and "measurement phase", some of the processing done by the informer affected the measurements and we got confusing benchmark results in https://github.com/kubernetes/kubernetes/pull/131168. This has been fixed together with some other issues (https://github.com/kubernetes/kubernetes/pull/133941, https://github.com/kubernetes/kubernetes/pull/134010) and now results make more sense (see below). However, although the conversion shows up in CPU profiles, it's not slow enough to have much impact.

The main reason for this change is that it enables extending the internal type with fields that get set by the [ResourceSlice tracker](https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker). The additional field is needed for the [revised device taint handling](https://github.com/kubernetes/enhancements/pull/5512) because it informs the eviction controller about which DeviceTaintRule status it needs to update. With this PR, dynamicresources plugin and allocator only use the internal type and get it either from the informer (this commit) or the ResourceSlice tracker.

Shifting the conversion may also enable other future optimizations because internalizing strings amortizes itself better compared to doing it repeatedly, but this has not been explored.

#### Which issue(s) this PR is related to:

Replaces: https://github.com/kubernetes/kubernetes/pull/131168

#### Special notes for your reviewer:

Performance comparison when using 36 cores on a 5 year old workstation (warning, not quite up-to-date):

[performance.md](https://github.com/user-attachments/files/22379336/performance.md)

Same for 8 cores shows slightly higher performance improvements:
```
numactl --physcpubind=0-7 make test-integration WHAT=./test/integration/scheduler_perf/dra/... KUBE_CACHE_MUTATION_DETECTOR=false KUBE_TIMEOUT=-timeout=1h ETCD_LOGLEVEL=warn KUBE_TEST_VMODULE="''" FULL_LOG=y ARTIFACTS=/tmp SHORT=-short=false KUBE_TEST_ARGS='-run=^$ -count=6 -benchtime=1ns -bench=.'
```

[performance-8-cores.md](https://github.com/user-attachments/files/22720597/performance-8-cores.md)

Interestingly, absolute performance is higher with less cores. Perhaps CPU cache effects?

TLDR: moderate performance improvements between 0 and 5%.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

pohly committed:
DRA ResourceSlice: convert to internal format in informer

The original idea was that avoiding repeated conversion of ResourceSlices to
the internal type by doing it in an informer would bring some performance
advantage. Because the DRA scheduler_perf benchmarks didn't clearly distinguish
between "setup phase" and "measurement phase", some of the processing done by
the informer affected the measurements and we got confusing benchmark results
in https://github.com/kubernetes/kubernetes/pull/131168. This has been fixed
together with some other
issues (https://github.com/kubernetes/kubernetes/pull/133941,
https://github.com/kubernetes/kubernetes/pull/134010) and now results make more
sense. However, although the conversion shows up in CPU profiles, it's not slow
enough to have much impact.

The main reason for this change is that it enables extending the internal type
with fields that get set by the ResourceSlice
tracker (staging/src/k8s.io/dynamic-resource-allocation/resourceslice/tracker). Those
additional fields are needed for the revised device taint
handling (https://github.com/kubernetes/enhancements/pull/5512) because that
change will remove the device taints that are embedded inside the Device
struct. Without that there will be no place where the tracker can report all
taints that apply to a device, a separate data structure and method of
communicating when it changes would be needed. With this commit,
dynamicresources plugin and allocator only use the internal type and get it
either from the informer (this commit) or the ResourceSlice tracker.

Shifting the conversion may also enable other future optimizations because
internalizing strings amortizes itself better compared to doing it repeatedly,
but this has not been explored.

pohly committed:
DRA resource slice tracker: unit test with only label change

This verifies that updating a slice with no relevant taint changes is stored
and emitted as event.

pohly committed:
DRA resource slice tracker: refactor change tracking

In syncSlice we can skip the t.patchedResourceSlices.Add if it
has been determined earlier that nothing changed in the slice.

Documentation and variables get changed for the sake of (subjective?) clarity.

pohly committed:
DRA E2E: cover ResourceSlice change

We had tests for "DeviceClass added/modified", but not for ResourceSlice
changes. This adds those.

These tests cover event handling in the scheduler. They could also be
integration tests, but as E2E tests they are fast and we get better testing in
different scenarios (= Prow jobs).

nojnhuh reviewed (approved):
Overall LGTM. Glad to see some more concrete results from this.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/uniquestring_test.go:
```diff
@@ -1,5 +1,5 @@
 /*
-Copyright 2025 The Kubernetes Authors.
+Copyright 2024 The Kubernetes Authors.
```
Updated.

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/structured/internal/allocatortesting/allocator_testing.go:
```diff
@@ -4998,6 +4999,18 @@ func TestAllocator(t *testing.T,
+	for i := range slices {
+		var slice draapi.ResourceSlice
+		if err := draapi.Convert_v1_ResourceSlice_To_api_ResourceSlice(slices[i], &slice, nil); err != nil {
+			panic(fmt.Errorf("convert ResourceSlice: %w", err))
```
Changed.

pohly commented:
> I assume there is no device taints intersection there?

That is correct. There's no reason to make that code more complex and/or make it depend on additional code (the converting informer).

nojnhuh reviewed (approved):
/lgtm

pohly commented on staging/src/k8s.io/dynamic-resource-allocation/api/informer.go:
```diff
@@ -0,0 +1,189 @@
+			// This must not get blocked when the consumer stops reading,
+			// hence the stopChan.
+			select {
+			case w.resultChan <- e:
```
Unfortunately this additional goroutines leaks quite a bit in test/integration/scheduler_perf:

```
    goleak.go:50: found unexpected goroutines:
        [Goroutine 158156 in state chan send, with k8s.io/dynamic-resource-allocation/api.(*watchResourceSlice).run on top of the stack:
        k8s.io/dynamic-resource-allocation/api.(*watchResourceSlice).run(0xc0079510b0, {{0x466e390?, 0xc0070ebf00?}, 0xc001a2b500?})
        	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/api/informer.go:184 +0x17b
        created by k8s.io/dynamic-resource-allocation/api.newResourceSliceInformer.func2 in goroutine 158039
        	/home/prow/go/src/k8s.io/kubernetes/staging/src/k8s.io/dynamic-resource-allocation/api/informer.go:107 +0x1e5
        ]
```

It looks like the informer helper got stuck on a write to its own result channel. That shouldn't happen, for several reasons:
- The reflector should consume all of those events.
- It calls Stop in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/client-go/tools/cache/reflector.go#L882, which in turn should cause `run` to give up on writing to the result channel - however, that's only under certain circumstances.

Need to investigate further...

/hold

pohly replied:
Duh! Preventing a blocking write was incomplete because I left the original write outside the `select` unchanged - each event got written twice!

https://github.com/kubernetes/kubernetes/compare/6b259cedb7c88eab8c0fe84c6a356c0e9ad55a4f..75190d9c66e44b3ed3427253f98182f10ea0c616

I'm still not entirely sure whether the context needs to be canceled explicitly (https://github.com/kubernetes/kubernetes/pull/134105/commits/d2b8647b61eeb890778733fd0e935d7bbb917261). I'm still investigating that:

```
diff --git a/staging/src/k8s.io/client-go/tools/cache/reflector.go b/staging/src/k8s.io/client-go/tools/cache/reflector.go
index 9b33e01ef2d..c81c5846107 100644
--- a/staging/src/k8s.io/client-go/tools/cache/reflector.go
+++ b/staging/src/k8s.io/client-go/tools/cache/reflector.go
@@ -359,7 +359,7 @@ func (r *Reflector) RunWithContext(ctx context.Context) {
                        r.watchErrorHandler(ctx, r, err)
                }
        }, r.backoffManager, true, ctx.Done())
-       logger.V(3).Info("Stopping reflector", "type", r.typeDescription, "resyncPeriod", r.resyncPeriod, "reflector", r.name)
+       logger.V(3).Info("Stopping reflector", "type", r.typeDescription, "resyncPeriod", r.resyncPeriod, "reflector", r.name, "cause", context.Cause(ctx))
 }
 
 var (
```

Hmm, that just says `cause="context canceled"`, so it's not the cancellation that I had added. I think it's safe to remove it again, it wasn't necessary.

/hold cancel

pohly replied:
Yes. I haven't added dedicated unit tests and they probably would have caught this subtle interaction with a reflector anyway, but this code is exercised.

pohly commented:
/test pull-kubernetes-integration

nojnhuh reviewed (approved):
/lgtm

pohly commented:
/assign @towca @macsko

macsko commented on pkg/controller/devicetainteviction/wrappers_test.go:
```diff
@@ -0,0 +1,82 @@
+}
+
+// Devices sets the devices field of the inner object.
+func (wrapper *ResourceSliceWrapper) Devices(names ...string) *ResourceSliceWrapper {
```
Is this method used outside the package?

pohly committed:
DRA internal API: add JSON tags

This is useful for klog.Format() and diff.Diff() which both use JSON to render
structs in a more readable manner. In particular the `omitempty` is important
to keep that output short and similar to the output of k8s.io/api types.

pohly committed:
DRA internal API: remove some dead code

These types were not actually used.

towca commented:
From the perspective of Cluster Autoscaler integration this change decreases readability (because instead of treating all the DRA types similarly, one is now special-cased) and introduces code churn (similar to what happens in this PR for k/k). These are not huge issues and I assume that alternative approaches would result in worse problems for the eviction controller + no performance benefits from fewer conversions? If so, this seems like a fair trade-off.

Other than that, I don't think the semantics of the integration actually change at all, right? The initial plan of plumbing the SliceTracker into CA and just using it instead of a real ResourceSlice lister when listing the slices should still work after we perform all the code churn first?

pohly commented:
> Other than that, I don't think the semantics of the integration actually change at all, right? The initial plan of plumbing the SliceTracker into CA and just using it instead of a real ResourceSlice lister when listing the slices should still work after we perform all the code churn first?

Yes. But you are right about the churn. Now comes the (for me) embarrassing part... :hot_face: 

I realized that the approach of informing the eviction controller about DeviceTaintRule taints through the ResourceSlice tracker doesn't work when ResourceSlices get deleted while workloads run. We explicitly allow that, I just hadn't considered this gap. I need to make the eviction controller handle rules itself, and then this PR is no longer needed for device taints.

It may still provide some performance benefit in the future if we convert to UniqueString more aggressively, but right now that doesn't justify this PR. Sorry for wasting your time! I'll close it for now, we can come back to it perhaps later when evaluating performance of some of the newer features (partitionable, mixins, consumable capacity -> @mortent, @sunya-ch ).

/close
